src/data.rs

```diff
@@ -530,6 +530,97 @@ impl<'a> DataTree<'a> {
         Ok(unsafe { DataTree::from_raw(context, rnode) })
     }
 
+    pub fn _parse_op(
+        &'a mut self,
+        data: impl AsRef<[u8]>,
+        format: DataFormat,
+        op_type: ffi::lyd_type::Type,
+    ) -> Result<DataNodeRef<'a, 'a>> {
```

Comment:
I'm confused about this method signature.

`lyd_parse_op()` has the following params related to data trees:
```
 * @param[in] parent Optional parent to connect the parsed nodes to.
 * @param[out] tree Optional full parsed data tree. If @p parent is set, set to NULL.
 * @param[out] op Optional pointer to the operation (action/RPC) node.
```

You're passing a NULL pointer to `parent`, which is the only input data tree (`tree` and `op` are output parameters). So why does this function need to take `&mut self`?

The why I'm seeing this, `lyd_parse_op` will allocate a new data tree to `op`, which is independent from `&mut self`, so the return value must be a `DataTree` that owns that data. Returning a `DataNodeRef` means `op` will leak once it goes out of scope.

Reply:
I believe you are correct. I actually am using the restconf routine in DataNodeRef and wrote these 4 functions for completeness sake to cover the other 3 transport based OP types. I'll redo these.